Implement all mix/max functions in a (hopefully) more optimization amendable way

WaffleLapkin opened the pull request:
Previously the graph was like this:

```
min -> Ord::min -> min_by -> match on compare() (in these cases compare = Ord::cmp)
                                      ^
                                      |
                                 min_by_key
```
now it looks like this:
```
min -> Ord::min -> `<=` <- min_by_key

min_by -> `Ordering::is_le` of `compare()`
```
(`max*` and `minmax*` are the exact same, i.e. they also use `<=` and `is_le`)

I'm not sure how to test this, but it should probably be easier for the backend to optimize.


r? @scottmcm
cc https://github.com/rust-lang/rust/issues/115939#issuecomment-2622161134

WaffleLapkin committed:
implement all min/max fns in terms of `<`/`is_lt`

`<` seems to be the "lucky one" for llvm

scottmcm reviewed (changes requested):
I think it's fine to not have a codegen test for this, though if you really wanted to you could probably make one with tuples, since those have `lt` and friends overridden separately.  It'd be fragile, though, since #133984 and the LLVM upgrade are probably going to change our output.

So I think a couple of updated doctests would be nice, then you're good.  I have some other comments, but those are all up to your discretion depending whether you think they're worth bothering.

WaffleLapkin commented on library/core/src/cmp.rs:
```diff
@@ -981,7 +981,7 @@ pub trait Ord: Eq + PartialOrd<Self> {
         Self: Sized,
     {
-        max_by(self, other, Ord::cmp)
+        if self <= other { other } else { self }
```
Sure, let's use our lucky op C:

WaffleLapkin replied:
Not gonna lie, using `<` for `min`/`max` is one of the most confusing options, since it requires doing `other < self`. But whatever, I did it and it works.

WaffleLapkin committed:
improve doc tests for (min/max/minmax).* functions

- add tests for `a == b` where missing
- try to make all the tests more similar
- try to use more illustrative test values

WaffleLapkin commented on library/core/src/cmp.rs:
```diff
@@ -1414,6 +1450,24 @@ pub macro PartialOrd($item:item) {
+///     fn cmp(&self, other: &Self) -> Ordering { Ordering::Equal }
+/// }
+///
+/// assert_eq!(cmp::min(Equal("v1"), Equal("v2")).0, "v1");
```
The values were chosen after the function's argument names, this feels confusing...

bors commented:
:pushpin: Commit c5835cd648da820fa4875376bb3fb15bb4511e0b has been approved by `scottmcm`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=scottmcm c5835cd648da820fa4875376bb3fb15bb4511e0b -->
<!-- homu: {"type":"Approved","sha":"c5835cd648da820fa4875376bb3fb15bb4511e0b","approver":"scottmcm","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

rust-timer commented:
Queued 20325765a05dc7b86aeadbd191aa4f4ad76407fb with parent e08cd3cf05e5bfa3323cc21ea8f81f4a15a2f969, future [comparison URL](https://perf.rust-lang.org/compare.html?start=e08cd3cf05e5bfa3323cc21ea8f81f4a15a2f969&end=20325765a05dc7b86aeadbd191aa4f4ad76407fb).
There is currently 1 preceding artifact in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.3 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (20325765a05dc7b86aeadbd191aa4f4ad76407fb): [comparison URL](https://perf.rust-lang.org/compare.html?start=e08cd3cf05e5bfa3323cc21ea8f81f4a15a2f969&end=20325765a05dc7b86aeadbd191aa4f4ad76407fb&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=e08cd3cf05e5bfa3323cc21ea8f81f4a15a2f969&end=20325765a05dc7b86aeadbd191aa4f4ad76407fb&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.1%  | [0.1%, 0.1%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 0.4%  | [0.3%, 0.4%]   | 2     |
| Improvements ✅ <br /> (primary)   | -0.4% | [-0.4%, -0.4%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.1% | [-0.4%, 0.1%]  | 2     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=e08cd3cf05e5bfa3323cc21ea8f81f4a15a2f969&end=20325765a05dc7b86aeadbd191aa4f4ad76407fb&stat=max-rss)
<details>
<summary>Results (primary 0.8%, secondary 2.3%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 4.7%  | [3.5%, 6.6%]   | 3     |
| Regressions ❌ <br /> (secondary)  | 3.2%  | [1.4%, 6.3%]   | 8     |
| Improvements ✅ <br /> (primary)   | -3.0% | [-4.4%, -1.2%] | 3     |
| Improvements ✅ <br /> (secondary) | -1.2% | [-1.3%, -1.2%] | 2     |
| All ❌✅ (primary)                 | 0.8%  | [-4.4%, 6.6%]  | 6     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=e08cd3cf05e5bfa3323cc21ea8f81f4a15a2f969&end=20325765a05dc7b86aeadbd191aa4f4ad76407fb&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=e08cd3cf05e5bfa3323cc21ea8f81f4a15a2f969&end=20325765a05dc7b86aeadbd191aa4f4ad76407fb&stat=size:linked_artifact)
<details>
<summary>Results (primary -0.0%, secondary -0.0%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.2%  | [0.0%, 0.5%]   | 15    |
| Regressions ❌ <br /> (secondary)  | 0.0%  | [0.0%, 0.0%]   | 18    |
| Improvements ✅ <br /> (primary)   | -0.1% | [-0.4%, -0.0%] | 36    |
| Improvements ✅ <br /> (secondary) | -0.4% | [-0.4%, -0.4%] | 3     |
| All ❌✅ (primary)                 | -0.0% | [-0.4%, 0.5%]  | 51    |
</details>

**Bootstrap**: 777.851s -> 778.832s (0.13%)
**Artifact size**: 328.84 MiB -> 328.84 MiB (0.00%)
